modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-ktor/infrastructure/ApiClient.kt.mustache

```diff
@@ -196,6 +197,7 @@ import {{packageName}}.auth.*
             this.method = requestConfig.method.httpMethod
             headers.filter { header -> !UNSAFE_HEADERS.contains(header.key) }.forEach { header -> this.header(header.key, header.value) }
             if (requestConfig.method in listOf(RequestMethod.PUT, RequestMethod.POST, RequestMethod.PATCH)) {
+                contentType(ContentType.Application.Json)
```

Comment:
instead of hard coding to applicaiton/json, what about updating api.mustache to use content-type defined in the sepc instead similar to what we've done in https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-okhttp/api.mustache#L222 ?

Reply:
@wing328 My first try was:
```
contentType(ContentType.parse("{{contentType}}"))
```

I also just tried:

```
contentType(ContentType.parse("{{{mediaType}}}"))
```

Both result in an empty value:
```
contentType(ContentType.parse(""))
```

Also, even if I get this correctly, I'm pretty sure we would need some conditionals to add the correct gradle dependencies in the template.

At this point I know very little about Mustache and the project layout overall.

Reply:
can you please share your spec? or a minimal spec to reproduce the issue?

> My first try was:

did you do it in api.mustache inside the `operation` mustache tag?

I would suggest plotting the fix in https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-okhttp/api.mustache#L222 to the api.mustache for `ktor-jvm` library.

Reply:
> did you do it in api.mustache inside the `operation` mustache tag?
I see I need to be in the scope of the `operation` now, but anyway, I cannot set the content type there due to the current design of the operation and the `jsonRequest` and `request` abstractions in `ApiClient`.

The current fix is done in a method called `request`, whose only caller is `jsonRequest`, so I highly doubt that "making this right" and set the dynamic Content Type will actually work.

I'll try a little bit longer, though.

Reply:
You can use this fix with customized template (e.g. -t option CLI) if you need this fix urgently for your use case.

Reply:
Yeah, I also tried to add a parameter in `RequestConfig`, but this class is not specific to ktor-jvm.

I could add a header in `RequestConfig.headers`, but again, ContentType is listed in `UNSAFE_HEADERS` and will be removed.

I see no easy way to fix the bug without a relatively major refactoring.

Reply:
If you look at https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-okhttp/api.mustache#L222, the content-type header is set there instead of ApiClient.

can we not do the same? or you saying jvm-okhttp has the issue for your use case?

Reply:
@wing328 is it time for a big redesign? should I give it a try?

Reply:
I don't think I fully understand the problem/limitation that requires the redesign.

Can you please PM me via Slack to further discuss this when you've time?

Reply:
> I don't think I fully understand the problem/limitation that requires the redesign.

It's not easy to understand the problem without looking at the code. Basically, the content type is being set in a super class method that has no access to the config, IIRC.

> Can you please PM me via Slack to further discuss this when you've time?

The issue ended up being fixed by https://github.com/OpenAPITools/openapi-generator/pull/16843, so I'll close this PR.

That solution ended up getting the content type from the header, or a hardcoded value as a fallback, BTW, so we still haven't "fixed" this properly as you suggested.